README
# -*- RD -*-

= README

== Ruby-GNOME2 -- Ruby bindings for GNOME-2.x

This is a set of bindings for the GNOME-2.x and GNOME-3.x libraries to
use from Ruby 1.9.x, 2.0.0 and 2.1.

== Release packages

=== ruby-gtk2

  * Ruby/GLib2:           GLib 2.12.x or later
  * Ruby/ATK:             ATK 1.12.x or later
  * Ruby/Pango:           Pango 1.14.x or later
  * Ruby/GdkPixbuf2:      GTK+ 2.0.x or later
  * Ruby/GTK2:            GTK+ 2.10.x or later
  * Ruby/GObjectIntrospection: GObject Introspection 1.32.1 or later

=== ruby-gtk3

  * ruby-gtk2 - Ruby/GTK2
  * Ruby/GDK3:            GTK+ 3.4.2 or later
  * Ruby/GTK3:            GTK+ 3.4.2 or later

=== ruby-gnome2-all

  * ruby-gtk2 + ruby-gtk3 - duplicated libraries
  * Ruby/RSVG:            librsvg 2.8.0 or later
  * Ruby/Poppler:         poppler-glib 0.8.0 or later
  * Ruby/VTE              VTE 0.12.1 or later
  * Ruby/GtkSourceView2:  GtkSourceView 2.0.0 or later
  * Ruby/GIO2:            GIO 2.16.x or later
  * Ruby/CairoGObject:    cairo-gobject 1.12.10 or later
  * Ruby/Clutter:         Clutter 1.12.0 or later
  * Ruby/ClutterGTK:      Clutter-GTK 1.2.0 or later
  * Ruby/ClutterGStreamer:Clutter-GStreamer 2.0.0 or later
  * Ruby/GtkSourceView3:  GtkSourceView 3.4.2 or later
  * Ruby/VTE3:            VTE 0.32.2 or later
  * Ruby/WebKitGTK:       WebKitGTK+ 1.8.1 or later (for GTK+ 3)
  * Ruby/WebKitGTK2:      WebKitGTK+ 1.8.1 or later (for GTK+ 2)
  * Ruby/GStreamer:       GStreamer 1.0.0 or later
  * Ruby/GooCanvas:       GooCanvas 0.8.0 or later

=== Experimental / Incomplete:

  * ...

=== Deprecated

  * Ruby/PanelApplet:     gnome-panel 2.6.0 or later (deprecated since 0.16.0)
  * Ruby/GConf2:          GConf 2.0.x or later
  * Ruby/GNOME2:          libgnome-2.0.x, libgnomeui-2.0.x or later
  * Ruby/GnomeCanvas2:    libgnomecanvas-2.0.x or later
  * Ruby/GnomePrint:      libgnomeprint-2.8.x or later
  * Ruby/GnomePrintUI:    libgnomeprintui-2.6.x or later
  * Ruby/GnomeVFS:        GnomeVFS 2.0.x or later
  * Ruby/GtkHtml2:        GtkHtml2 2.0.x or later
  * Ruby/GtkGLExt:        GtkGLExt 1.0.3 or later
  * Ruby/Libart2:         libart_lgpl 2.3.12 or later
  * Ruby/GtkSourceView:   GtkSourceView 1.0.1 or later
  * Ruby/Libglade2:       Libglade 2.0.x or later
  * Ruby/GtkMozEmbed:     GtkMozEmbed (Mozilla 1.7.x or FireFox-1.0.x or later)

== Install

  % ruby extconf.rb
  % make
  % sudo make install

To compile and install a particular sub-binding, you can add arguments:

  % ruby extconf.rb [subdir]...
    e.g.) ruby extconf.rb glib2 pango atk gdk_pixbuf2 gtk2

Or you can compile each sub-binding:

  % cd <each sub-directory>
  % ruby extconf.rb
  % make
  % sudo make install

=== extconf.rb options

: --ruby
   ruby directory

: --topsrcdir
   top source directory

: --topdir
    top directory

: --strict
    if some libraries fail to compile/install, "make"
    command returns 1(exit 1)

=== Environment Variables

: CAIRO_PATH
    rcairo path (for Win32).

=== Dependencies

You should install a sub-binding with its dependencies.  The
current dependencies are:

: Ruby/GLib
   none.

: Ruby/ATK
   depends on Ruby/GLib.

: Ruby/Pango
   depends on Ruby/GLib, rcairo(optional).

: Ruby/GdkPixbuf
   depends on Ruby/GLib.

: Ruby/GTK
   depends on Ruby/GLib, Ruby/ATK, Ruby/Pango,
   Ruby/GdkPixbuf, rcairo(optional)

: Ruby/RSVG
   depends on Ruby/GLib, Ruby/GdkPixbuf2.

: Ruby/Poppler
   depends on Ruby/GLib, Ruby/GdkPixbuf(optional), rcairo(optional).

: Ruby/VTE
   depends on Ruby/GTK.

: Ruby/GtkSourceView2
   depends on Ruby/GTK.

: Ruby/GIO2
   depends on Ruby/GLib.

=== Experimental

: Ruby/GStreamer
   depends on Ruby/GLib.

: Ruby/GooCanvas
   depends on Ruby/GTK.

=== Deprecated

: Ruby/PanelApplet
   depends on Ruby/GTK (since libpanel-applet 2.25) or
   Ruby/GNOME (before libpanel-applet 2.25).

: Ruby/Libart
   depends on Ruby/GLib(Use mkmf-gnome2.rb only).

: Ruby/GConf
   depends on Ruby/GLib.

: Ruby/GNOME
   depends on Ruby/GnomeCanvas.

: Ruby/GnomeCanvas
   depends on Ruby/GTK, Ruby/Libart.

: Ruby/GnomeVFS
   depends on Ruby/GLib.

: Ruby/GnomePrint
   depends on Ruby/GLib, Ruby/Pango, Ruby/Libart2.

: Ruby/GnomePrintUI
   depends on Ruby/GTK, Ruby/GnomePrint.

: Ruby/GtkHtml2
   depends on Ruby/GTK.

: Ruby/GtkGLExt
   depends on Ruby/GTK, rbogl.

: Ruby/GtkSourceView
   depends on Ruby/GTK.

: Ruby/Libglade
   depends on Ruby/GTK, Ruby/GNOME(optional), REXML(optional).

== Bugs

Please report bugs either in our bug tracker
((<URL: http://sourceforge.net/tracker/?group_id=53614&atid=470969 >))

or on the ruby-gnome2-devel-en / ruby-gnome2-devel-ja mailing list
((<URL: http://sourceforge.net/mail/?group_id=53614 >))

== Copying

Copyright (c) 2002-2014 Ruby-GNOME2 Project Team

This program is free software.
You can distribute/modify this program under the terms of
the GNU LESSER GENERAL PUBLIC LICENSE Version 2.1.

== Project Website

((<URL: http://ruby-gnome2.sourceforge.jp/ >))


NEWS
# -*- RD -*-

= NEWS

== Ruby-GNOME 3.5.1: 2021-01-17

This is a release for Windows.

=== Changes

==== All

  * windows: Add workaround for mingw-w64-x86_64-gettext-0.21-1 or
    later. Dummy (({DllMain()})) is defined.

==== Ruby/Pango

  * Fixes

    * Fixed a bug that can't be started.
      [GitHub#1456][Reported by Akira Ouchi]

==== Ruby/GObjectIntrospection

  * Fixes

    * Fixed a bug that (({NoMethodError})) is raised on invalid
      signature for constructor.

=== Thanks

  * Akira Ouchi

== Ruby-GNOME 3.5.0: 2021-01-11

This is a release that adds support for Ractor.

Ruby/GObjectIntrospection has some backward incompatibilities for
Ractor support. If you have any problem, please report it to
https://github.com/ruby-gnome/ruby-gnome/issues .

=== Changes

==== Ruby/GLib2

  * Improvements

    * Added support for Ractor.

    * Added support for signal handlers and virtual methods in
      included module.

    * Added support for (({try_convert})) protocol for property setter.

    * Added support for converting tuple (({GVariant})) to Ruby.

==== Ruby/GIO2

  * Improvements

    * Added support for GIO 2.70.

    * Added (({Gio::RubyInputStream})) to use Ruby objects as
      (({Gio::InputStream})).

    * Added (({Gio::RubyOutputStream})) to use Ruby objects as
      (({Gio::OutputStream})).

==== Ruby/GObjectIntrospection

  * Improvements

    * Added support for Ractor. This introduced some backward
      incompatiblities.

    * Added support for "transfer full" for out (({GError})).
      [GitHub#1437][Reported by mcclumpherty]

    * Added support for changing whether GVL is unlocked per object by
      the following APIs.

      * (({GObjectIntrospection::FunctionInfo#set_lock_gvl_default}))

      * (({GObjectIntrospection::FunctionInfo#add_lock_gvl_predicate}))

      * (({GObjectIntrospection::Loader#prepare_function_info_lock_gvl}))

    * Added support for converting from raw argument to enum.

    * Added support for (({GList<GVariant>})) return value.

    * Added support for reporting an error in callback.

  * Fixes

    * Fixed a bug that virtual functions of grandparent class can't be
      implemented.
      [GitHub#1433][Patch by shibafu]

==== Ruby/Pango

  * Improvements

    * Added support for Ruby 3.2.

==== Ruby/GTK3

  * Improvements

    * Improved documentation.
      [GitHub#1454][Patch by Andy Maleh]

    * Added support for Ruby 3.2.

==== Ruby/GDK4

  * Improvements

    * Updated pkg-config ID.
      [GitHub#1435][Patch by Sasi Olin]

==== Ruby/GTK4

  * Improvements

    * Removed needless rsvg2 dependency on Windows.
      [GitHub#1440][Reported by HuBandiT]

  * Fixes

    * Fixed typos in warning messages.
      [GitHub#1442][Patch by HuBandiT]
      [GitHub#1415][Reported by rubyFeedback]

=== Thanks

  * shibafu

  * Sasi Olin

  * mcclumpherty

  * HuBandiT

  * rubyFeedback

  * Andy Maleh

== Ruby-GNOME 3.4.9: 2021-08-10

This is a bug fix release of 3.4.8.

=== Changes

==== Ruby/GObjectIntrospection

  * Fixes

    * Fixed a bug that (({gpointer})) to Ruby conversion breaks a
      value when pointer value is (({2 ** 32})) or larger.

== Ruby-GNOME 3.4.8: 2021-08-09

This is a bug fix release of 3.4.7.

=== Changes

==== Ruby/GObjectIntrospection

  * Fixes

    * Fixed a bug that (({gpointer})) to Ruby conversion breaks a
      value when pointer value is (({2 ** 32})) or larger.

== Ruby-GNOME 3.4.7: 2021-07-30

This is a release that improves virtual function support.

=== Changes

==== Ruby/GLib2

  * Improvements

    * (({GLib::Error})): Added support for setting (({code})) and
      (({domain})) automatically.

==== Ruby/GObjectIntrospection

  * Improvements

    * Added support for returning object from callback.

  * Fixes

    * Fixed a bug that (({GError})) detection doesn't work.

== Ruby-GNOME 3.4.6: 2021-07-17

This is a bug fix release for macOS.

=== Changes

==== Ruby/Pango

  * Fixes

    * Fixed a bug that (({require "pango"})) is failed on environment
      that has multiple font types.
      [GitHub#1429][Reported by Cameron Gose]

=== Thanks

  * Cameron Gose

== Ruby-GNOME 3.4.5: 2021-07-07

This is a release that supports implementing virtual functions in
Ruby.

=== Changes

==== Ruby/GLib2

  * Improvements

    * Added (({GError})) domain and code for Ruby.

==== Ruby/GObjectIntrospection

  * Improvements

    * (({GObjectIntrospection::BaseInfo#container})): Added.

    * (({GObjectIntrospection::ObjectInfo#class_struct})): Added.

    * (({GObjectIntrospection::StructInfo#find_field})): Added.

    * (({RVAL2GI_VFUNC_INFO()})): Added.

    * Added support for implementing virtual functions in Ruby.
      [GitHub#1386][Based on patch by Yuto Tokunaga]

      You need to define (({virtual_do_#{virtual_function_name}}))
      method in (({type_register}))-ed class.

    * Added support for implementing virtual functions of interface in
      Ruby.
      [GitHub#985][Reported by Matijs van Zuijlen]
      [GitHub#1938][Reported by Yuto Tokunaga]

    * Added support for "transfer everything" UTF-8 return/output
      value.

    * Changed to accepted one character for (({gunichar})).
      [GitHub#1426][Reported by rubyFeedback]

    * (({GObjectIntrospection::CallableInfo#can_throw_gerror?})): Added.

    * Added support for (({GError **})) in callback.

    * Added support for returning (({GList<GObject *>})) from callback.

    * Changed to return (({[]})) for (({NULL})) list.

  * Fixes

    * Fixed a bug that wrong type information is used for output
      arguments.

==== Ruby/Pango

  * Added support for (({PangoFT2})).

  * Added support for (({PangoFc})).

  * Added support for (({PangoOT})).

  * Added support for (({PangoCairoFontMaps})).

  * Updated gem metadata.
    [GitHub#1428][Patch by Gabriel Mazetto]

=== Thanks

  * Yuto Tokunaga

  * Matijs van Zuijlen

  * rubyFeedback

  * Gabriel Mazetto

== Ruby-GNOME 3.4.4: 2021-04-22

This is a bug fix release for Windows.

=== Changes


extconf.rb
=begin
  top-level extconf.rb for Ruby-GNOME2

  $Id: extconf.rb,v 1.17 2007/10/22 12:19:17 ktou Exp $

  Copyright (C) 2003-2005 Ruby-GNOME2 Project Team
=end

require 'English'
require 'mkmf'
require 'fileutils'
require 'pathname'

priorlibs = [
  "glib2",
  "cairo-gobject",
  "gobject-introspection",
  "gio2",
  "gdk_pixbuf2",
  "pango",
  "atk",
  "gdk3",
  "gtk3",
]

unsupported_libraries = [
]

#
# detect sub-directories
#
$ruby = File.join(RbConfig::CONFIG['bindir'], RbConfig::CONFIG['RUBY_INSTALL_NAME'] + RbConfig::CONFIG['EXEEXT'])
$ruby = arg_config("--ruby", $ruby)

rm = "rm -f"
if /mswin/ =~ RUBY_PLATFORM
  rm = "del"
end


$srcdir = File.dirname(__FILE__)
$topsrcdir = $configure_args["--topsrcdir"] ||= $srcdir
$topdir = $configure_args["--topdir"] ||= Dir.pwd
$strict = $configure_args["--strict"] ? "--strict" : ""

$srcdir = File.expand_path($srcdir)
$topsrcdir = File.expand_path($topsrcdir)
$topdir = File.expand_path($topdir)

subdirs = ARGV.select{|v|  /^--/ !~ v}

if subdirs.empty?
  subdirs = Dir.glob($topsrcdir+"/*/extconf.rb")
  subdirs.collect! do |subdir|
    subdir[0..$topsrcdir.size] = ""
    File.dirname(subdir)
  end
  priorlibs &= subdirs
  subdirs -= priorlibs
  subdirs = priorlibs + subdirs #Change the order
end
subdirs = subdirs.delete_if do |subdir|
  subdir.end_with?("-no-gi")
end
subdirs -= unsupported_libraries

#
# generate sub-directory Makefiles
#
target_modules = []
ignore_modules = []

ruby, *ruby_args = Shellwords.shellwords($ruby)
if ARGV.grep(/\A--ruby=/)
  extra_args = ["--ruby=#{$ruby}"] + ARGV.reject {|arg| /\A--ruby=/ =~ arg}
else
  extra_args = ARGV.dup
end

subdirs.each do |subdir|
  STDERR.puts("#{$0}: Entering directory `#{subdir}'")
  FileUtils.mkdir_p(subdir)
  topdir = File.join(*([".."] * subdir.split(/\/+/).size))
  dir = $topsrcdir
  dir = File.join(topdir, dir) unless Pathname.new(dir).absolute?
  srcdir = File.join(dir, subdir)
  args = ruby_args + ["-C", subdir, File.join(srcdir, "extconf.rb"),
                      "--topsrcdir=#{dir}", "--topdir=#{topdir}",
                      *extra_args]
  ret = system(ruby, *args)
  STDERR.puts("#{$0}: Leaving directory '#{subdir}'")
  if ret
    target_modules << subdir
  else
    ignore_modules << subdir
  end
end
puts "\n-----"
unless target_modules.empty?
  puts "Target libraries: #{target_modules.join(', ')}"
end
unless ignore_modules.empty?
  puts "Ignored libraries: #{ignore_modules.join(', ')}"
end

#
# generate top-level Makefile
#

def run_make_in_sub_dir(sub_dir, target)
  if /mswin/ =~ RUBY_PLATFORM
    "	$(COMMAND) '#{sub_dir}' $(MAKE) #{target}"
  else
    "	(cd '#{sub_dir}' && $(MAKE) #{target})"
  end
end


File.open("Makefile", "w") do |makefile|
  makefile.print(<<-EOM)
TOPSRCDIR = #{$topsrcdir}
COMMAND = #{$ruby} #{$topsrcdir}/exec_make.rb #{$strict}
RM = #{rm}
EOM

  ["all", "install", "site-install", "clean", "distclean"].each do |target|
    makefile.print(<<-EOM)
#{target}:
EOM
    target_modules.each do |target_module|
      sub_target = "#{target}-#{target_module}"
      makefile.print(<<-EOM)
#{target}: #{sub_target}
#{sub_target}:
	#{run_make_in_sub_dir(target_module, target)}

EOM
    end
  end

  makefile.print(<<-EOM)
distclean: distclean-toplevel
distclean-toplevel:
	$(RM) Makefile mkmf.log
EOM
end

puts "-----"
puts "Done."

$makefile_created = true
